src/ClusterManager/virtual_cluster_status.py

```diff
@@ -0,0 +1,205 @@
+                        pod_status.get("preemptable_memory", Memory()).to_dict(),
+                    "preemptable_gpu":
+                        pod_status.get("preemptable_gpu", Gpu()).to_dict(),
+                })
+                vc_preemptable_used[vc_name] += pod_preemptable_res
+
+            # Account all jobs without pods in vc
+            jobs_without_pods = vc_jobs_without_pods.get(vc_name, [])
+            for job in jobs_without_pods:
+                job_params = job["jobParams"]
+                job_res_params = get_resource_params_from_job_params(job_params)
+                job_res = ClusterResource(params=job_res_params)
+
+                if not job["preemptionAllowed"]:
+                    vc_used[vc_name] += job_res
+                else:
+                    vc_preemptable_used[vc_name] += job_res
+                logger.info("Added job %s resource %s to the usage of vc %s",
+                            job, job_res, vc_name)
+
+        return vc_used, vc_preemptable_used
+
+    def __get_cluster_resource_count(self):
+        cluster = self.cluster_status
+        capacity = ClusterResource(
+            params={
+                "cpu": cluster.cpu_capacity,
+                "memory": cluster.memory_capacity,
+                "gpu": cluster.gpu_capacity,
+            }
+        )
+        avail = ClusterResource(
+            params={
+                "cpu": cluster.cpu_available,
+                "memory": cluster.memory_available,
+                "gpu": cluster.gpu_available,
+            }
+        )
+        reserved = ClusterResource(
+            params={
+                "cpu": cluster.cpu_reserved,
+                "memory": cluster.memory_reserved,
+                "gpu": cluster.gpu_reserved,
+            }
+        )
+        return capacity, avail, reserved
+
+
+class VirtualClusterStatusesFactory(object):
+    def __init__(self, cluster_status, vc_list):
+        self.cluster_status = cluster_status
+        self.vc_info = get_vc_info(vc_list)
+
+    def make(self):
+        try:
+            vc_pod_statuses = self.__get_vc_pod_statuses()
+            vc_jobs = self.__get_vc_jobs()
+
+            vc_statuses = {
+                vc_name: VirtualClusterStatus(
+                    vc_name,
+                    self.vc_info,
+                    self.cluster_status,
+                    self.cluster_status.node_statuses,
+                    vc_pod_statuses,
+                    vc_jobs
+                ) for vc_name in self.vc_info
+            }
+        except:
+            logger.exception("Failed to make vc statuses", exc_info=True)
+            vc_statuses = None
+
+        return vc_statuses
+
+    def __get_vc_pod_statuses(self):
+        pod_statuses = self.cluster_status.pod_statuses
+        vc_pod_statuses = {}
+        for vc_name in self.vc_info:
+            vc_pod_statuses[vc_name] = {
+                name: pod_status
+                for name, pod_status in pod_statuses.items()
```

Comment:
Maybe put this for loop to the outer loop? This essentially traverse the pod_statuses len(vc_inf) times